lib/services/pages.js

```diff
@@ -243,16 +243,24 @@ function publish(uri, data, locals) {
     site = getSite(prefix, locals),
     timeoutLimit = timeoutConstant * timeoutPublishCoefficient;
 
-  return getPublishData(uri, data).then(function (pageData) {
-    const published = 'published',
-      publicUrl = pageData.url,
-      componentList = references.getPageReferences(pageData);
-
-    if (!references.isUrl(publicUrl)) {
-      throw new Error('Client: Page must have valid url to publish.');
-    }
+  return getPublishData(uri, data)
+    .then(function (pageData) {
+      if ( _.isArray(site.publishingChain) ) {
```

Comment:
"loop through the array of publishing functions, finding the first one that resolves"

"this is a variant of the Chain of Responsibility pattern, which we're calling Reject Quickly/Resolve Slowly. functions should reject quickly if the data doesn't match the format they want, and _afterwards_ resolve the generated url if the data _does_ match"

etc